File: frontend/lib/administration/user-management-policy.ts
import { SUPER_ADMIN_ROLE } from "@/types/permissions";

export const ADMIN_ROLE = "Admin";

const ADMIN_MANAGED_ROLES = new Set(["Engineer", "Operator", "Viewer"]);

const SUPER_ADMIN_CREATABLE_ROLES = new Set([
  SUPER_ADMIN_ROLE,
  ADMIN_ROLE,
  "Engineer",
  "Operator",
  "Viewer",
]);

const ADMIN_CREATABLE_ROLES = ADMIN_MANAGED_ROLES;

export function filterVisibleUsers<T extends { role: string }>(
  users: T[],
  actorRole: string,
): T[] {
  if (actorRole === SUPER_ADMIN_ROLE) {
    return users;
  }

  if (actorRole === ADMIN_ROLE) {
    return users.filter((user) => user.role !== SUPER_ADMIN_ROLE);
  }

  return [];
}

export function canManageUser(actorRole: string, targetRole: string): boolean {
  if (actorRole === SUPER_ADMIN_ROLE) {
    return true;
  }

  if (actorRole === ADMIN_ROLE) {
    return ADMIN_MANAGED_ROLES.has(targetRole);
  }

  return false;
}

export function getCreatableRoles(actorRole: string): string[] {
  if (actorRole === SUPER_ADMIN_ROLE) {
    return [...SUPER_ADMIN_CREATABLE_ROLES];
  }

  if (actorRole === ADMIN_ROLE) {
    return [...ADMIN_CREATABLE_ROLES];
  }

  return [];
}

export function canAssignRole(actorRole: string, role: string): boolean {
  return getCreatableRoles(actorRole).includes(role);
}

File: backend/app/core/authorization/user_management_policy.py
from app.core.authorization.roles import SUPER_ADMIN_ROLE

ADMIN_ROLE = "Admin"

ADMIN_MANAGED_ROLES: frozenset[str] = frozenset({"Engineer", "Operator", "Viewer"})

SUPER_ADMIN_CREATABLE_ROLES: frozenset[str] = frozenset(
    {SUPER_ADMIN_ROLE, ADMIN_ROLE, "Engineer", "Operator", "Viewer"},
)

ADMIN_CREATABLE_ROLES: frozenset[str] = ADMIN_MANAGED_ROLES


def creatable_roles_for(actor_role: str) -> frozenset[str]:
    if actor_role == SUPER_ADMIN_ROLE:
        return SUPER_ADMIN_CREATABLE_ROLES
    if actor_role == ADMIN_ROLE:
        return ADMIN_CREATABLE_ROLES
    return frozenset()


def can_manage_user(actor_role: str, target_role: str) -> bool:
    if actor_role == SUPER_ADMIN_ROLE:
        return True
    if actor_role == ADMIN_ROLE:
        return target_role in ADMIN_MANAGED_ROLES
    return False


def can_assign_role(actor_role: str, new_role: str) -> bool:
    return new_role in creatable_roles_for(actor_role)
